framework/src/mediaplayer/mediaplayer.c

```diff
@@ -0,0 +1,530 @@
+{
+	if (m->state == MEDIA_STATE_PLAYING) {
+		m->state = MEDIA_STATE_PAUSING;
+		return MEDIA_OK;
+	} else {
+		return -MEDIA_ERROR;
+	}
+}
+
+int media_resume(media_t *m)
+{
+	if (m->state == MEDIA_STATE_PAUSED) {
+		m->state = MEDIA_STATE_PLAYING;
+		sem_post(&g_sem_playing);
+		return MEDIA_OK;
+	} else {
+		return -MEDIA_ERROR;
+	}
+}
+
+void media_setvol(media_t *m, unsigned char vol)
+{
+	return;
+}
+
+bool media_isplaying(media_t *m)
+{
+	return (m->state == MEDIA_STATE_PLAYING);
+}
+
+unsigned char media_getvol(media_t *m)
+{
+	return MEDIA_OK;
+}
+
+int media_seek(media_t *m, int time_ms)
+{
+	return MEDIA_OK;
+}
+
+int media_setposition(media_t *m, int time_ms)
+{
+	return MEDIA_OK;
+}
+
+int media_getposition(media_t *m)
+{
+	return MEDIA_OK;
+}
+
+/***
+ * Functions run on threads
+ */
+
+static int read_wav_header(media_t *m)
+{
+	//struct pcm_config config;
+	struct wav_header_s header;
+	int ret;
+
+	ret = read(m->fd, &header, sizeof(header));
+	if (ret != sizeof(header)) {
+		return -MEDIA_ERROR_UNKNOWN_FILE;
+	}
+
+	// Todo: Add header information to media_t
+
+	return MEDIA_OK;
+}
+
+media_t *media_open(char *path, media_op_e op, media_type_e type)
+{
+	media_t *m = (media_t *)malloc(sizeof(media_t));
+	if (m == NULL) {
+		return NULL;
+	}
+
+	m->fd = -1;
+	m->op = op;
+	m->state = MEDIA_STATE_CREATED;
+	m->type = type;
+
+	if (type == MEDIA_TYPE_WAV || type == MEDIA_TYPE_PCM) {
+		if (op == MEDIA_OP_PLAYBACK) {
+			m->fd = open(path, O_RDONLY);
+			if (m->fd < 0) {
+				return NULL;
+			}
+
+			if (m->type == MEDIA_TYPE_WAV) {
+				read_wav_header(m);
+			}
+			m->state = MEDIA_STATE_PLAYING;
+
+		} else if (op == MEDIA_OP_RECORD) {
+			m->fd = open(path, O_RDWR | O_CREAT | O_TRUNC);
+			if (m->fd < 0) {
+				return NULL;
+			}
+			m->state = MEDIA_STATE_RECORDING;
+		}
+	} else if (type == MEDIA_TYPE_WAV_STREAM) {
+		// Todo: Supports network streaming
+		return NULL;
+	}
+
+	return m;
+}
+
+void media_close(media_t *m)
+{
+	if (m == NULL) {
+		return;
+	}
+	if (m->fd >= 0) {
+		close(m->fd);
+	}
+	free(m);
+	return;
+}
+
+
+int on_media_state_playing(struct pcm *pcmout, media_t *m, char *buffer, unsigned int buffer_size)
+{
+	int readed;
+	int remain;
+	int ret;
+	readed = read(m->fd, buffer, buffer_size);
+	if (readed) {
+		remain = pcm_bytes_to_frames(pcmout, readed);
+		while (remain > 0) {
+			ret = pcm_writei(pcmout, buffer + readed - pcm_frames_to_bytes(pcmout, remain), remain);
+			if (ret <= 0) {
+				return -MEDIA_ERROR_PCM_WRITE;
+			} else {
+				remain -= ret;
+			}
+		}
+	}
+
+	return readed;
+}
+
+int on_media_state_recording(struct pcm *pcmin, media_t *m, char *buffer, unsigned int buffer_size)
+{
+	int readed;
+	int remain;
+	int ret;
+
+	readed = pcm_readi(pcmin, buffer, pcm_bytes_to_frames(pcmin, buffer_size));
+	if (readed) {
+		remain = pcm_frames_to_bytes(pcmin, readed);
+		while (remain > 0) {
+			ret = write(m->fd, buffer + pcm_frames_to_bytes(pcmin, readed) - remain, remain);
```

Comment:
This is a media player, writing on file is application's role.
If application doesn't write data on file system & send recorded data through the network each of defined time.
How it can handle?
I think callback function should be added as a parameter.
int on_media_state_recording(struct pcm *pcmin, media_t *media, char *buffer, record_cb callback)
{ 
    pcm_readi();
    if (buffer is full) {
       callback;
   }
}

Reply:
or you can find another good way instead of callback.

Reply:
I think we can keep writing to file in this mediaplayer code. it is one of our use cases we thought initially. 
The STREAM flag of media type is for this purpose only. The buffer would be passed to the application for processing. But we are yet to discuss the method through which this should done , either callback or some other way. We are actually not clear of all the use cases.